crypto/evp/evp_enc.c

```diff
@@ -336,7 +302,25 @@ int EVP_EncryptUpdate(EVP_CIPHER_CTX *ctx, unsigned char *out, int *outl,
             return 0;
         }
     }
+
+    /* Past this point inl has to be in bytes. */
     i = ctx->buf_len;
+    /*
+     * Aliasing in and out in this way is counter intuitive and only supported
+     * for legacy applications.
+     * In fact the check below excludes just one special case of partially
+     * overlapping buffers, while the docs say we return an error if in and out
+     * are partially overlapping.
+     * It is quite possible that the check below will be hardened in the future
+     * to return an error whenever there is any overlapping between in and out
+     * buffers.
```

Comment:
I apologize, but I still find this confusing/contradicting. It gives an impression that there is something specific legacy application *may* rely on, something that works and is currently supported [and something we consider stopping supporting]. But simple fact of life is that **partially** overlapping buffers were never supported. Yes, something might have worked, but it was never actually supported. To draw an analogy, you can call memcpy with **partially** overlapping buffers, and it will do something. And this something might even work out for you that time. But does it make it supported for legacy applications? No. Well, one can argue that memcpy is off the hook, because there is disclaimer in its manual page. But does lack of disclaimer in earlier OpenSSL documentation put an obligation on us to actually support it? I'd say no, not really(*). As it is **naturally** assumed limitation. It's just that originally there was no check and application code was held fully responsible for possible mishaps, and now there is. One can argue that current check is buggy as it doesn't catch all corner cases, but you can't really say that its hardening would conflict with legacy applications, as they were never to actually rely on its lack to start with. In other words if anything, this note should simply state that check can be hardened in the future. [Though it kind of calls for following question. If we recognize that check is not 100%, then why isn't it fixed now. One can argue that it's a trade-off decision, to catch most obvious cases, but fast. As previous check.]

(*) BTW, if answer was yes, then solution would have be different and more convoluted.

[Just in case for bypasser's reference, **fully** overlapping buffers are considered different from **partially** overlapping buffers, and are supported. Well, one can argue that there is ambiguity, because output from block cipher is padded, and padded output won't be fully overlapping with input. Yes, but padding is performed in separate finalization call, when application input is fully consumed.]

Reply:
Well, I am actually **not** sure if there is any real application doing that.
There are however are some tests doing exactly that.

If it would be my decision, I would drop the admittedly hard to explain
is_pointer_offset_by_bias-thing completely and simplify this check to:

`if (is_any_overlapping(out, inl + (bl - 1), in, inl)) {`

... and adjust the test accordingly.
If that would be acceptable the comment could be dropped as well.